src/matter/common/MessageExchange.ts

```diff
@@ -146,12 +159,25 @@ export class MessageExchange<ContextT> {
                     throw new Error(`Incorrect ack received. Expected ${sentMessageIdToAck}, received: ${ackedMessageId}`);
                 }
             } else {
-                // The other side has received our previous message
-                this.sentMessageAckSuccess?.();
-                this.retransmissionTimer?.stop();
-                this.sentMessageAckSuccess = undefined;
-                this.sentMessageAckFailure = undefined;
-                this.sentMessageToAck = undefined;
+                // If we only expect an Ack without data but got data, reject with this
+                if (this.expectAckOnly && !SecureChannelProtocol.isStandaloneAck(protocolId, messageType)) {
```

Comment:
This leaks higher layer concerns into the lower layer and it will be really good to avoid this.
You should walk me through what is the intent since this change makes the code very hard to follow.
The MessageExchange should only ensure valid communication at the exchange level, if the flow is not valid for a specific message flow, this should be handled by the higher level controlling that message flow.

Reply:
Yes I know but I did not found a better way to do that.
The topic is relevant for cases where a response is normally supressed - then the normal answer is an Standalone Ack. Our current logic will not forward this Standalone Ack to the queue, so there is not way for the "upper layers" to see that, which is ok because the promise gets resolved.

But now the special case comes: if an error occurs the answer is NOT a standalone ack but a StatusResponse message. So the current logic will resolve the promise (as before) and will send the message to the Queue - but the upper layer has no chance to know that there was an "unexpected StatusResponse" as response. SO yes we could so a "getNextMessage" but we need a special logic because we should not wait for any new we just need to check if there is something "already received" (and hope it is ours - no idea if there could be edge cases).

I tried several ideas and all ended in meehh code and this was the best one.